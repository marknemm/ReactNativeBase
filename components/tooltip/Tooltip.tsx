import { useControlledState } from '@hooks/state-hooks';
import { Tooltip as RneuiTooltip } from '@rneui/themed';
import { useCallback } from 'react';
import type { TooltipProps } from './Tooltip.interfaces';

/**
 * A tooltip component.
 *
 * The visibility state can be controlled internally or externally.
 *
 * @param props The component {@link TooltipProps}.
 * @returns The {@link Tooltip} component.
 */
const Tooltip: React.FC<TooltipProps> = ({
  backgroundColor = 'gray',
  children,
  onClose,
  onOpen,
  visible,
  ...tooltipProps
}) => {
  const [tooltipVisible, setTooltipVisible] = useControlledState(visible, { initValue: false });

  const onCloseCb = useCallback(() => {
    setTooltipVisible(false);
    onClose?.();
  }, [onClose, setTooltipVisible]);

  const onOpenCb = useCallback(() => {
    setTooltipVisible(true);
    onOpen?.();
  }, [onOpen, setTooltipVisible]);

  return (
    <RneuiTooltip
      backgroundColor={backgroundColor}
      onClose={onCloseCb}
      onOpen={onOpenCb}
      visible={tooltipVisible}
      {...tooltipProps}
    >
      { children }
    </RneuiTooltip>
  );
};

export type * from './Tooltip.interfaces';
export default Tooltip;
