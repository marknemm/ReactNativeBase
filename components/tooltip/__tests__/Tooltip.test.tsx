import Tooltip from '@components/tooltip/Tooltip';
import { Text } from '@rneui/themed';
import AppProvider from '@test/contexts/app/AppProvider';
import { render, screen, userEvent } from '@testing-library/react-native';
import { View } from 'react-native';

describe('<Tooltip />', () => {
  const childrenStr = 'Tooltip content';
  const popoverStr = 'Popover content';
  const popover = (
    <View testID="rnb-test-popover">
      <Text>{popoverStr}</Text>
    </View>
  );

  describe('controlled visibility state', () => {
    it('shows popover on visible prop change', () => {
      render(
        <Tooltip
          popover={popover}
          visible
        >
          <Text>{childrenStr}</Text>
        </Tooltip>,
        { wrapper: AppProvider }
      );

      const popoverText = screen.getByText(popoverStr);
      expect(popoverText).toBeVisible();
    });

    it('hides popover on visible prop change', () => {
      render(
        <Tooltip
          popover={popover}
          visible
        >
          <Text>{childrenStr}</Text>
        </Tooltip>,
        { wrapper: AppProvider }
      );

      let popoverText = screen.getByText(popoverStr);
      expect(popoverText).toBeVisible();

      screen.rerender(
        <Tooltip
          popover={popover}
          visible={false}
        >
          <Text>{childrenStr}</Text>
        </Tooltip>
      );

      popoverText = screen.queryByText(popoverStr);
      expect(popoverText).toBeFalsy();
    });

    it('calls onClose and onOpen callbacks but does not auto open/close', async () => {
      const onClose = jest.fn();
      const onOpen = jest.fn();
      render(
        <Tooltip
          onClose={onClose}
          onOpen={onOpen}
          popover={popover}
          visible={false}
        >
          <Text>{childrenStr}</Text>
        </Tooltip>,
        { wrapper: AppProvider }
      );

      const text = screen.getByText(childrenStr);
      await userEvent.press(text);
      let popoverText = screen.queryByText(popoverStr);

      expect(onOpen).toHaveBeenCalledTimes(1);
      expect(popoverText).toBeFalsy();

      screen.rerender(
        <Tooltip
          onClose={onClose}
          onOpen={onOpen}
          popover={popover}
          visible
        >
          <Text>{childrenStr}</Text>
        </Tooltip>
      );

      popoverText = screen.getByText(popoverStr);
      await userEvent.press(popoverText);
      popoverText = screen.getByText(popoverStr);

      expect(onClose).toHaveBeenCalledTimes(1);
      expect(popoverText).toBeVisible();
    });
  });

  describe('uncontrolled (auto) visibility state', () => {
    it('does not show popover by default', () => {
      render(
        <Tooltip popover={popover}>
          <Text>{childrenStr}</Text>
        </Tooltip>,
        { wrapper: AppProvider }
      );

      const popoverText = screen.queryByText(popoverStr);

      expect(popoverText).toBeFalsy();
    });

    it('shows popover on press', async () => {
      const onOpen = jest.fn();
      render(
        <Tooltip
          onOpen={onOpen}
          popover={popover}
        >
          <Text>{childrenStr}</Text>
        </Tooltip>,
        { wrapper: AppProvider }
      );

      const text = screen.getByText(childrenStr);
      await userEvent.press(text);

      expect(onOpen).toHaveBeenCalledTimes(1);

      const popoverText = screen.getByText(popoverStr);
      expect(popoverText).toBeVisible();
    });

    it('hides popover on popover press', async () => {
      const onOpen = jest.fn();
      const onClose = jest.fn();
      render(
        <Tooltip
          onClose={onClose}
          onOpen={onOpen}
          popover={popover}
        >
          <Text>{childrenStr}</Text>
        </Tooltip>,
        { wrapper: AppProvider }
      );

      const text = screen.getByText(childrenStr);
      await userEvent.press(text);

      expect(onOpen).toHaveBeenCalledTimes(1);

      let popoverText = screen.getByText(popoverStr);
      expect(popoverText).toBeVisible();

      await userEvent.press(popoverText);

      expect(onClose).toHaveBeenCalledTimes(1);

      popoverText = screen.queryByText(popoverStr);
      expect(popoverText).toBeFalsy();
    });
  });

  describe('snapshots', () => {
    it('renders correctly', () => {
      render(
        <Tooltip />,
        { wrapper: AppProvider }
      );

      expect(screen.toJSON()).toMatchSnapshot();
    });

    it('renders correctly with text content', () => {
      render(
        <Tooltip>
          <Text>{childrenStr}</Text>
        </Tooltip>,
        { wrapper: AppProvider }
      );

      expect(screen.toJSON()).toMatchSnapshot();
    });

    it('renders correctly with hidden popover', () => {
      render(
        <Tooltip popover={popover}>
          <Text>{childrenStr}</Text>
        </Tooltip>,
        { wrapper: AppProvider }
      );

      expect(screen.toJSON()).toMatchSnapshot();
    });

    it('renders correctly with visible popover', () => {
      render(
        <Tooltip
          popover={popover}
          visible
        >
          <Text>{childrenStr}</Text>
        </Tooltip>,
        { wrapper: AppProvider }
      );

      expect(screen.toJSON()).toMatchSnapshot();
    });
  });
});
