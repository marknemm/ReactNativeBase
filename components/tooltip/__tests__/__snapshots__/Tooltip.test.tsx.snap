// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Tooltip /> snapshots renders correctly 1`] = `
<View
  collapsable={false}
>
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
  />
</View>
`;

exports[`<Tooltip /> snapshots renders correctly with hidden popover 1`] = `
<View
  collapsable={false}
>
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
  >
    <Text
      accessibilityRole="text"
      style={
        {
          "color": "#242424",
          "fontFamily": undefined,
          "fontSize": 23.57142857142857,
          "fontStyle": "normal",
          "fontWeight": "normal",
          "marginBottom": 0,
          "textAlign": "left",
          "textDecorationLine": "none",
          "textDecorationStyle": "solid",
        }
      }
    >
      Tooltip content
    </Text>
  </View>
</View>
`;

exports[`<Tooltip /> snapshots renders correctly with text content 1`] = `
<View
  collapsable={false}
>
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
  >
    <Text
      accessibilityRole="text"
      style={
        {
          "color": "#242424",
          "fontFamily": undefined,
          "fontSize": 23.57142857142857,
          "fontStyle": "normal",
          "fontWeight": "normal",
          "marginBottom": 0,
          "textAlign": "left",
          "textDecorationLine": "none",
          "textDecorationStyle": "solid",
        }
      }
    >
      Tooltip content
    </Text>
  </View>
</View>
`;

exports[`<Tooltip /> snapshots renders correctly with visible popover 1`] = `
<View
  collapsable={false}
>
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
  >
    <Text
      accessibilityRole="text"
      style={
        {
          "color": "#242424",
          "fontFamily": undefined,
          "fontSize": 23.57142857142857,
          "fontStyle": "normal",
          "fontWeight": "normal",
          "marginBottom": 0,
          "textAlign": "left",
          "textDecorationLine": "none",
          "textDecorationStyle": "solid",
        }
      }
    >
      Tooltip content
    </Text>
  </View>
  <Modal
    animationType="fade"
    hardwareAccelerated={false}
    onShow={[Function]}
    transparent={true}
    visible={true}
  >
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "backgroundColor": "#fafafa14",
          "flex": 1,
          "opacity": 1,
        }
      }
    >
      <View>
        <View
          style={
            {
              "backgroundColor": "transparent",
              "height": 0,
              "left": 0,
              "overflow": "visible",
              "position": "absolute",
              "top": 0,
              "width": 0,
            }
          }
        >
          <Text
            accessibilityRole="text"
            style={
              {
                "color": "#242424",
                "fontFamily": undefined,
                "fontSize": 23.57142857142857,
                "fontStyle": "normal",
                "fontWeight": "normal",
                "marginBottom": 0,
                "textAlign": "left",
                "textDecorationLine": "none",
                "textDecorationStyle": "solid",
              }
            }
          >
            Tooltip content
          </Text>
        </View>
        <View
          style={
            {
              "left": -7.5,
              "position": "absolute",
              "top": -2,
            }
          }
        >
          <View
            style={
              {
                "backgroundColor": "transparent",
                "borderBottomColor": "gray",
                "borderBottomWidth": 15,
                "borderLeftColor": "transparent",
                "borderLeftWidth": 8,
                "borderRightColor": "transparent",
                "borderRightWidth": 8,
                "borderStyle": "solid",
                "height": 0,
                "width": 0,
              }
            }
            testID="RNE__Tooltip_Triangle"
          />
        </View>
        <View
          style={
            {
              "alignItems": "center",
              "backgroundColor": "gray",
              "borderRadius": 10,
              "display": "flex",
              "flex": 1,
              "height": 40,
              "justifyContent": "center",
              "left": -17.999,
              "padding": 10,
              "position": "absolute",
              "top": 10,
              "width": 150,
            }
          }
          testID="tooltipPopoverContainer"
        >
          <View
            testID="rnb-test-popover"
          >
            <Text
              accessibilityRole="text"
              style={
                {
                  "color": "#242424",
                  "fontFamily": undefined,
                  "fontSize": 23.57142857142857,
                  "fontStyle": "normal",
                  "fontWeight": "normal",
                  "marginBottom": 0,
                  "textAlign": "left",
                  "textDecorationLine": "none",
                  "textDecorationStyle": "solid",
                }
              }
            >
              Popover content
            </Text>
          </View>
        </View>
      </View>
    </View>
  </Modal>
</View>
`;
